### PR TITLE
feat(object): enable versionning on bucket resource

### PR DESCRIPTION
**Pull request by remyleone**

### CODE REVIEW

File: scaleway/helpers_object.go [hunk 1]
```diff
@@ -16,6 +20,9 @@ func newS3Client(region, accessKey, secretKey string) (*s3.S3, error) {
 	config.WithRegion(region)
 	config.WithCredentials(credentials.NewStaticCredentials(accessKey, secretKey, ""))
 	config.WithEndpoint("https://s3." + region + ".scw.cloud")
+	if os.Getenv("S3_DEBUG") != "" {
```
Comment (Sh4d1) [inline scaleway/helpers_object.go:23]: should we rather use `SCW_DEBUG` ?

Comment (remyleone) [reply scaleway/helpers_object.go:23]: It seems more user friendly yes, even if they are two different SDKs

File: scaleway/helpers_object.go [hunk 2]
```diff
@@ -16,6 +20,9 @@ func newS3Client(region, accessKey, secretKey string) (*s3.S3, error) {
 	config.WithRegion(region)
 	config.WithCredentials(credentials.NewStaticCredentials(accessKey, secretKey, ""))
 	config.WithEndpoint("https://s3." + region + ".scw.cloud")
+	if os.Getenv(scw.DebugEnv) != "" {
```
Comment (jerome-quere) [inline scaleway/helpers_object.go:23]: Replace by env(`TF_LOG`) == strtolower("debug")

File: scaleway/resource_object_bucket_test.go [hunk 3]
```diff
@@ -142,3 +144,119 @@ func testSweepStorageObjectBucket(_ string) error {
...
+		rs, ok := s.RootModule().Resources[n]
+		if !ok {
+			return fmt.Errorf("not found: %s", n)
+		}
+
+		if rs.Primary.ID == "" {
+			return fmt.Errorf("no ID is set")
+		}
+
+		conn := tt.Meta.s3Client
```
Comment (jerome-quere) [inline scaleway/resource_object_bucket_test.go:222]: use `newS3ClientFromMeta`

File: scaleway/provider.go [hunk 4]
```diff
@@ -131,6 +132,7 @@ func Provider(config *ProviderConfig) plugin.ProviderFunc {
 type Meta struct {
 	// scwClient is the Scaleway SDK client.
 	scwClient *scw.Client
+	s3Client  *s3.S3
```
Comment (jerome-quere) [inline scaleway/provider.go:135]: Remove

File: scaleway/provider.go [hunk 5]
```diff
@@ -180,8 +182,17 @@ func buildMeta(config *MetaConfig) (*Meta, error) {
 		return nil, err
 	}
 
+	region, _ := scwClient.GetDefaultRegion()
```
Comment (jerome-quere) [inline scaleway/provider.go:185]: remove

File: scaleway/resource_object_bucket.go [hunk 6]
```diff
@@ -140,6 +156,18 @@ func resourceScalewayObjectBucketRead(ctx context.Context, d *schema.ResourceDat
 
 	_ = d.Set("endpoint", objectBucketEndpointURL(bucketName, region))
 
+	// Read the versioning configuration
+	versioningResponse, err := s3Client.GetBucketVersioning(&s3.GetBucketVersioningInput{
+		Bucket: aws.String(bucketName),
```
Comment (jerome-quere) [inline scaleway/resource_object_bucket.go:161]: Add context

File: scaleway/resource_object_bucket.go [hunk 7]
```diff
@@ -140,6 +156,18 @@ func resourceScalewayObjectBucketRead(ctx context.Context, d *schema.ResourceDat
...
+	})
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	vcl := flattenObjectBucketVersioning(versioningResponse)
+	if err := d.Set("versioning", vcl); err != nil {
+		return diag.FromErr(fmt.Errorf("error setting versioning: %s", err))
+	}
```
Comment (jerome-quere) [inline scaleway/resource_object_bucket.go:169]: _ := d.Set("versioning",  flattenObjectBucketVersioning(versioningResponse))

File: scaleway/resource_object_bucket.go [hunk 8]
```diff
@@ -194,3 +228,34 @@ func resourceScalewayObjectBucketDelete(ctx context.Context, d *schema.ResourceD
...
+		c := v[0].(map[string]interface{})
+
+		if c["enabled"].(bool) {
+			vc.Status = aws.String(s3.BucketVersioningStatusEnabled)
+		} else {
+			vc.Status = aws.String(s3.BucketVersioningStatusSuspended)
+		}
+	} else {
+		vc.Status = aws.String(s3.BucketVersioningStatusSuspended)
+	}
```
Comment (jerome-quere) [inline scaleway/resource_object_bucket.go:247]: create an expandBucketVersioning function

File: scaleway/resource_object_bucket.go [hunk 9]
```diff
@@ -194,3 +228,34 @@ func resourceScalewayObjectBucketDelete(ctx context.Context, d *schema.ResourceD
...
+
+	i := &s3.PutBucketVersioningInput{
+		Bucket:                  aws.String(bucketName),
+		VersioningConfiguration: vc,
```
Comment (jerome-quere) [inline scaleway/resource_object_bucket.go:251]: add context

File: scaleway/resource_object_bucket.go [hunk 10]
```diff
@@ -194,3 +228,34 @@ func resourceScalewayObjectBucketDelete(ctx context.Context, d *schema.ResourceD
...
+	}
+	log.Printf("[DEBUG] S3 put bucket versioning: %#v", i)
```
Comment (jerome-quere) [inline scaleway/resource_object_bucket.go:253]: use `l.Debug("S3.....`

File: scaleway/resource_object_bucket_test.go [hunk 11]
```diff
@@ -142,3 +144,119 @@ func testSweepStorageObjectBucket(_ string) error {
 		return nil
 	})
 }
+
+func TestAccScalewayObjectBucket_Versioning(t *testing.T) {
+	tt := NewTestTools(t)
+	defer tt.Cleanup()
+
+	bucketName := acctest.RandomWithPrefix("tf-test-bucket")
```
Comment (jerome-quere) [inline scaleway/resource_object_bucket_test.go:152]: No random when using k7

**Review by jerome-quere — CHANGES_REQUESTED** [review]

File: scaleway/helpers_object.go (hunk 1 above)
Comment (remyleone) [reply scaleway/helpers_object.go]: We went with `TF_LOG==debug`

**Review by jerome-quere — APPROVED** [review]

**Pull request merged**